contracts/impls/FlowMarginProtocol2.sol

```diff
@@ -236,13 +237,13 @@ contract FlowMarginProtocol2 is FlowProtocolBase {
      * @param _baseTokenAmount The base token amount to withdraw.
      */
     function withdraw(LiquidityPoolInterface _pool, uint256 _baseTokenAmount) public nonReentrant poolIsVerified(_pool) {
-        require(getFreeBalance(_pool, msg.sender) >= int256(_baseTokenAmount), "Not enough free balance to withdraw!");
+        uint256 iTokenAmount = moneyMarket.redeemBaseTokenTo(msg.sender, _baseTokenAmount);
+        require(getFreeBalance(_pool, msg.sender) >= int256(iTokenAmount), "Not enough free balance to withdraw!");
```

Comment:
better to do require first

I think it saves gas on bad input?

Reply:
This way actually saves gas as the iTokenAmount doesn't have to be computed twice. If a transaction is reverted via require, it doesn't cost any gas anyways, no matter if the require is at the beginning or end.